packages/hadron-document/src/object-generator.ts

```diff
@@ -313,14 +327,29 @@ export class ObjectGenerator {
         (element.isAdded() || element.hasChangedKey()) &&
         element.currentKey !== ''
       ) {
-        // When a new field is added, check if that field
-        // was already added in the background.
-        originalFields.push({
-          path: [String(element.currentKey)],
-          value: DoesNotExist,
-        });
+        // When a new field is added, check if the original value of that
+        // field (which is typically that it was missing) was changed in
+        // the background. If there *was* another field in its place,
+        // that means that it was removed, and is added to `originalValue`
+        // at another point.
+        let wasRenamedToKeyOfPreviouslyExistingElement = false;
+        for (const otherElement of target.elements ?? []) {
+          if (
+            otherElement !== element &&
+            otherElement.key === element.currentKey
+          ) {
+            wasRenamedToKeyOfPreviouslyExistingElement = true;
+            break;
+          }
+        }
+        if (!wasRenamedToKeyOfPreviouslyExistingElement) {
+          originalFields.push({
+            path: [{ key: String(element.currentKey), isArrayIndex }],
+            value: DoesNotExist,
+          });
+        }
```

Comment:
This is kind of an independent issue, but didn’t appear in tests so far because when building the query in the non-dots-and-dollars mode, duplicate entries in `originalFields` would not cause trouble. For example, if property `a` with value `42` was removed and then later another element renamed to `a`, `originalFields` would contain two entries for `a`, one claiming that it does not exist, one claiming that it had the value `42`; the query would still (accidentally correctly) end up being `{ a: 42 }`, since the property in the query doc was simply overwritten.

In dots-and-dollars mode, however, the different values in `originalFields` are turned into multiple `$eq` checks which are then `$and`-ed together, so obviously there would be a conflict between an `$eq` asserting that the field did not exist and an `$eq` asserting that the field had its previous value.